common/src/main/java/org/opensearch/security/common/resources/ResourceAccessHandler.java

```diff
@@ -0,0 +1,549 @@
+import org.opensearch.security.spi.resources.exceptions.ResourceSharingException;
+import org.opensearch.security.spi.resources.sharing.Recipient;
+import org.opensearch.security.spi.resources.sharing.RecipientType;
+import org.opensearch.security.spi.resources.sharing.RecipientTypeRegistry;
+import org.opensearch.security.spi.resources.sharing.ResourceSharing;
+import org.opensearch.security.spi.resources.sharing.ShareWith;
+import org.opensearch.security.spi.resources.sharing.SharedWithScope;
+import org.opensearch.threadpool.ThreadPool;
+
+/**
+ * This class handles resource access permissions for users, roles and backend-roles.
+ * It provides methods to check if a user has permission to access a resource
+ * based on the resource sharing configuration.
+ *
+ * @opensearch.experimental
+ */
+public class ResourceAccessHandler {
+    private static final Logger LOGGER = LogManager.getLogger(ResourceAccessHandler.class);
+
+    private final ThreadContext threadContext;
+    private final ResourceSharingIndexHandler resourceSharingIndexHandler;
+    private final AdminDNs adminDNs;
+
+    public ResourceAccessHandler(
+        final ThreadPool threadPool,
+        final ResourceSharingIndexHandler resourceSharingIndexHandler,
+        AdminDNs adminDns
+    ) {
+        this.threadContext = threadPool.getThreadContext();
+        this.resourceSharingIndexHandler = resourceSharingIndexHandler;
+        this.adminDNs = adminDns;
+    }
+
+    /**
+     * Initializes the recipient types for users, roles, and backend roles.
+     * These recipient types are used to identify the types of recipients for resource sharing.
+     */
+    public void initializeRecipientTypes() {
+        RecipientTypeRegistry.registerRecipientType(Recipient.USERS.getName(), new RecipientType(Recipient.USERS.getName()));
+        RecipientTypeRegistry.registerRecipientType(Recipient.ROLES.getName(), new RecipientType(Recipient.ROLES.getName()));
+        RecipientTypeRegistry.registerRecipientType(
+            Recipient.BACKEND_ROLES.getName(),
+            new RecipientType(Recipient.BACKEND_ROLES.getName())
+        );
+    }
+
+    /**
+     * Returns a set of accessible resource IDs for the current user within the specified resource index.
+     *
+     * @param resourceIndex The resource index to check for accessible resources.
+     * @param listener      The listener to be notified with the set of accessible resource IDs.
+     */
+    public void getAccessibleResourceIdsForCurrentUser(String resourceIndex, ActionListener<Set<String>> listener) {
+        final UserSubjectImpl userSubject = (UserSubjectImpl) threadContext.getPersistent(
+            ConfigConstants.OPENDISTRO_SECURITY_AUTHENTICATED_USER
+        );
+        final User user = (userSubject == null) ? null : userSubject.getUser();
+
+        // If no user is authenticated, return an empty set
+        if (user == null) {
+            LOGGER.warn("Unable to fetch user details. User is null.");
```

Comment:
A resource sharing plugin indirectly cosumes the code via the `client` dependency, right?

Will this be also the case when security is disabled?

If so, what's the strategy for the case that security is disabled? Here it looks like that resources are not available at all when security is disabled (i.e., the user object is null).

Reply:
When security is disabled, the call will not reach here. Instead it would fail [here in the client] (https://github.com/opensearch-project/security/blob/e6d43b606608617667e5823aaf7ba3a022f56cec/client/src/main/java/org/opensearch/security/client/resources/ResourceSharingNodeClient.java#L68) with action not found

Reply:
Does that mean that resource plugins cannot be used when security is disabled?